Components: Standardize reduced motion handling using media queries

SainathPoojary opened the pull request:
Part of: https://github.com/WordPress/gutenberg/issues/68282


## What?
Refactors animation and transition styles to use `@media not (prefers-reduced-motion)` ensuring better accessibility for users who prefer reduced motion.

## Why?
Currently, many parts of the codebase do not consider users’ motion preferences, which may not be suitable for those with reduced motion settings. This PR addresses and resolves that issue.

## Screenshots or screencast <!-- if applicable -->

### Button:
Transition:

https://github.com/user-attachments/assets/bfa2f65c-2cc1-4fa2-ba92-7c3c7ffac81b

Animation

https://github.com/user-attachments/assets/cd5d45c6-fcc9-446c-8cdf-3b123958d835

### checkbox-control:
Transition:

https://github.com/user-attachments/assets/17e74c91-3927-4c8d-9ade-24241a60dbbf

### circular-option-picker:
Transition:

https://github.com/user-attachments/assets/d3cf13b7-27b7-409d-8100-5b6a7de4eee7

### form-toggle:
Transition:

https://github.com/user-attachments/assets/9240e760-72f6-498d-8b6b-c19540dc82fb

### form-token-field:
Transition:

https://github.com/user-attachments/assets/e48cec89-f433-4313-a5c5-64b45f36e010

### panel:
Transition:

https://github.com/user-attachments/assets/3f366b0e-9821-4834-b6b0-72973491dee8

### placeholder:
Transition:

https://github.com/user-attachments/assets/697c1b62-f957-4674-91fe-0a1faa552845

### resizable-box:
Transition:

https://github.com/user-attachments/assets/4ec8e2a8-a244-4d58-bf30-b838a74f1990

Animation:

https://github.com/user-attachments/assets/a1d1953a-1c1d-41dc-bed3-a4d539aac1b9

### tab-panel:
Transition:

https://github.com/user-attachments/assets/7ac84f73-be15-4750-a3b9-bc7f3b15ec2e

### drop-zone:
Transition:

https://github.com/user-attachments/assets/f48d2224-4d3d-4c91-a0f0-569586dac854

### animate
Animation:

https://github.com/user-attachments/assets/b5b5c56f-27c7-452d-a4e5-c77b1d2e1c2c

### modal
Animation:

https://github.com/user-attachments/assets/f67cfd6a-e0e4-49fb-ad42-e333c7ed74a5

### toolbar
Animation:

https://github.com/user-attachments/assets/9ddd1cff-60f8-449c-a7a4-5e1ac94d699f

SainathPoojary committed:
Button: Standardize reduced motion handling using media queries

SainathPoojary committed:
Button: Standardize animation reduced motion handling using media queries

SainathPoojary committed:
CheckboxControl: Standardize reduced motion handling using media queries

SainathPoojary committed:
CircularOptionPicker: Standardize reduced motion handling using media queries

SainathPoojary committed:
FormToggle: Standardize reduced motion handling using media queries

SainathPoojary committed:
FormTokenField: Standardize reduced motion handling using media queries

SainathPoojary committed:
Panel: Standardize reduced motion handling using media queries

SainathPoojary committed:
Placeholder: Standardize reduced motion handling using media queries

SainathPoojary commented:
Hey @t-hamano, 

This package includes several components. Would it be better to raise a new PR for the remaining components, or should I add them to this PR?

Components currently in PR:
```
button
checbox-control
circular-option-picker
form-toggle
form-token-field
panel
placeholder
```

Remaining components:
```
resizable-box
tab-panel
animate
modal
toolbar
drop-zone
```

t-hamano commented:
@SainathPoojary Thanks for working on this issue.

> This package includes several components. Would it be better to raise a new PR for the remaining components, or should I add them to this PR?

It will be easy to test this PR because almost all components are published in Storybook, so I personally think it's okay to include all components in this PR.

SainathPoojary committed:
ResizableBox: Standardize reduced motion handling using media queries

SainathPoojary commented:
Got it, I agree! I’ll include the remaining components in this PR.

SainathPoojary committed:
TabPanel: Standardize reduced motion handling using media queries

SainathPoojary committed:
DropZoneComponent: Standardize reduced motion handling using media queries

SainathPoojary committed:
Animate: Standardize reduced motion handling using media queries

SainathPoojary committed:
Modal: Standardize reduced motion handling using media queries

SainathPoojary committed:
Toolbar: Standardize reduced motion handling using media queries

SainathPoojary commented:
Hey @t-hamano,

I have standardized the transition and animation for the Component package and attached a video demonstrating the tests for each component. Please take a look when you have a moment.

Thank you!

SainathPoojary committed:
Animate: Fix stuck animation for slide-in animation

t-hamano commented on packages/components/src/circular-option-picker/style.scss:
```diff
@@ -35,9 +35,12 @@ $color-palette-circle-spacing: 12px;
+		transition: 100ms transform ease;
+	}
+
 	will-change: transform;
```
I think we could move the `will-change` inside the media query.

t-hamano commented on packages/components/src/resizable-box/style.scss:
```diff
@@ -60,8 +60,11 @@ $resize-handler-container-size: $resize-handler-size + ($grid-unit-05 * 2); // M
-	transition: transform 0.1s ease-in;
-	@include reduce-motion("transition");
+
+	@media not (prefers-reduced-motion) {
+		transition: transform 0.1s ease-in;
+	}
+
 	will-change: transform;
```
```suggestion
	@media not (prefers-reduced-motion) {
		transition: transform 0.1s ease-in;
		will-change: transform;
	}
```

t-hamano commented on packages/components/src/toolbar/toolbar/style.scss:
```diff
@@ -56,9 +56,10 @@
-			animation: components-button__appear-animation 0.1s ease;
-			animation-fill-mode: forwards;
-			@include reduce-motion("animation");
+			@media not (prefers-reduced-motion) {
+				animation: components-button__appear-animation 0.1s ease;
+				animation-fill-mode: forwards;
+			}
```
Do you know what this animation is intended to do? As far as I've tested, the outline animation when I focus on a button doesn't work.

Maybe we can improve or fix this in a follow-up.

SainathPoojary replied:
Hey @t-hamano, to better understand what this animation is doing, I increased the animation duration to 0.5s. Here’s the result.

https://github.com/user-attachments/assets/638db6c7-0f90-49fc-af39-d21e01e7183d

t-hamano replied:
Thank you! I didn't know there was an animation applied to it.

t-hamano reviewed (commented):
Thanks for the PR! I've tested all components on Storybook and everything appears to work as before.

In order for this PR to be shippable, could you confirm the following?

- Rebase the PR. This should pass all E2E tests (See https://github.com/WordPress/gutenberg/pull/68484).
- Update the [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md). Add a `Internal` section under the `Unreleased` section and add this PR as an item in it.

@WordPress/gutenberg-components If you have any feedback on this PR, please let us know 🙏

---

Note: The CI error reported in [the comment](https://github.com/WordPress/gutenberg/pull/68419#issuecomment-2567550864) (Storybook build error) does not occur in this PR.

Presumably the error only occurs when surrounded by a breakpoint mixin like below, but there are no media queries surrounded by a breakpoint mixin in this PR:

```scss
.selector {
	@include break-medium {
		@media not (prefers-reduced-motion) {
		}
	}
}
```

SainathPoojary committed:
Components: Move will-change inside media query for transform optimization

- ResizableBox: Optimized transform.
- CircularOptionPicker: Optimized transform.

SainathPoojary committed:
Components: Update CHANGELOG for reduced motion standardization

SainathPoojary committed:
Merge branch 'trunk' into update/components-reduce-motion-media-queries

SainathPoojary commented:
Hey @t-hamano,

I have implemented all the requested changes and also updated the code with the trunk, Please have a look at your convenience.

Thank you!

t-hamano reviewed (approved):
LGTM! All components appear to be correctly standardized.

Before shipping this PR, can you check the following points?

- Move the changelog entry to the correct place. It needs to be in the `Unreleased` section:
  ```markdown
  ## Unreleased

  ### Internal

  -   `Components`: Standardize reduced motion handling using media queries ([#68421](https://github.com/WordPress/gutenberg/pull/68421)).

  ## 29.1.0 (2025-01-02)
  ```
- Rebase this PR so that All CIs pass

SainathPoojary committed:
Components: Move CHANGELOG entry to Unreleased section

SainathPoojary commented:
Hey @t-hamano,

I have moved the changelog entry to the correct place now. Previously, I merged trunk into my PR, and I believe all the CI checks are now passing. Would you like me do it again?

Thank you!

t-hamano reviewed (changes requested):
Unfortunately, after testing this PR again, it appears that Storybook is no longer able to build. This is the same issue reported [here](https://github.com/WordPress/gutenberg/pull/68419#issuecomment-2567550864).

I don't know the exact reason yet, but it might be a good idea to merge the Storybook build's CI (#68466) first while we investigate the cause.

t-hamano commented:
I was wrong. After running `npm install` on another old branch, I didn't run `npm install` after switching to this branch 😅

In this PR, The StoryBook build is indeed successful. Just to be sure, I tried rebasing this PR on the trunk branch, and it worked just as well.
